sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs

```diff
@@ -0,0 +1,110 @@
+using Microsoft.IdentityModel.Tokens;
+
+namespace Azure.Messaging.WebPubSub
+{
+    /// <summary>
+    /// Web PubSub Authentication Policy.
+    /// </summary>
+    internal partial class WebPubSubAuthenticationPolicy : HttpPipelineSynchronousPolicy
+    {
+        private const int MaxTokenLength = 4096;
+
+        /// <summary>
+        /// Generates client bearwer token.
+        /// </summary>
+        /// <param name="audience"></param>
+        /// <param name="claims"></param>
+        /// <param name="key"></param>
+        /// <param name="expireAfter"></param>
+        /// <param name="hmacSha512">SHA512 if true, otherwise SHA256</param>
+        /// <returns></returns>
+        public static string GenerateAccessToken(
+            string audience,
+            IEnumerable<Claim> claims,
+            AzureKeyCredential key,
+            TimeSpan expireAfter = default,
+            bool hmacSha512 = false)
+        {
+            if (expireAfter == default) expireAfter = TimeSpan.FromMinutes(10);
+
+            var jwtToken = JwtUtils.GenerateJwtBearer(
+                audience: audience,
+                claims: claims,
+                expiresAfter: expireAfter,
+                key: key,
+                hmacSha512: hmacSha512
+            );
+
+            Debug.Assert(jwtToken.Length <= MaxTokenLength);
+
+            return jwtToken;
+        }
+
+        private static class JwtUtils
+        {
+            private static readonly JwtSecurityTokenHandler JwtTokenHandler = new JwtSecurityTokenHandler();
+            private static readonly ConditionalWeakTable<AzureKeyCredential, UniqueKey> KeyForCredential = new ConditionalWeakTable<AzureKeyCredential, UniqueKey>();
+
+            public static string GenerateJwtBearer(
+                string audience,
+                IEnumerable<Claim> claims,
+                TimeSpan expiresAfter,
+                AzureKeyCredential key,
+                string issuer = null,
+                bool hmacSha512 = false)
+            {
+                var expiresAt = DateTime.UtcNow.Add(expiresAfter);
+
+                var subject = claims == null ? null : new ClaimsIdentity(claims);
+                SigningCredentials credentials = null;
+                if (key != null)
+                {
+                    // Refer: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/releases/tag/5.5.0
+                    // From version 5.5.0, SignatureProvider caching is turned On by default, assign KeyId to enable correct cache for same SigningKey
+                    var securityKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(key.Key))
+                    {
+                        KeyId = KeyForCredential.GetOrCreateValue(key).Key
+                    };
+
+                    var algorithm = hmacSha512 ? SecurityAlgorithms.HmacSha512 : SecurityAlgorithms.HmacSha256;
+                    credentials = new SigningCredentials(securityKey, algorithm);
```

Comment:
Do we need to make sure we have enough key material for the specified algorithm? The [documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.tokens.signingcredentials.-ctor?view=azure-dotnet#Microsoft_IdentityModel_Tokens_SigningCredentials__ctor_Microsoft_IdentityModel_Tokens_SecurityKey_System_String_) didn't mention doing this check when the `SigningCredentials` is constructed. Perhaps it's not a problem if the service only generates 64 byte keys, but if not what happens if I use a 32 byte key and specify `hmacSha512=true`?